features/net/network-socket/nsapi_types.h

```diff
@@ -148,6 +148,18 @@ typedef enum nsapi_option {
 } nsapi_option_t;
 
 
+/*  C linkage definition
+ *
+ *  Attribute garuntees emitted function has C linkage.
+ *  Available in both C and C++
+ */
+#ifdef __cplusplus
+#define NSAPI_C_LINKAGE extern "C"
```

Comment:
It's used in nsapi_dns.h ([here](https://github.com/ARMmbed/mbed-os/pull/2497/files#diff-be743f4afff365499e8204bfac300f92R94)) which is a header and can be included from both C and C++